File: LeetCode/LeetCode Practice Problems/98- Validate Binary Search Tree.cpp
class Solution {
public:
    
    bool valid(TreeNode* root,long mn,long mx) {
        if(root == NULL) {
            return true;
        }
        
        if(root->val <= mn || root->val >= mx) {
            return false;
        }
        
        return valid(root->left , mn , root->val) && valid(root->right , root->val , mx);
        
    }
    
    bool isValidBST(TreeNode* root) {
        return valid(root,LONG_MIN,LONG_MAX);
    }
};

File: LeetCode/LeetCode Practice Problems/98-Validate Binary Search Tree.java
class Solution {
    
    public boolean validateBST(TreeNode root,long mn,long mx) {
        if(root == null) return true;
        if(root.val <= mn || root.val >= mx) {
            return false;
        }
        return validateBST(root.left , mn , root.val) && 
                validateBST(root.right , root.val , mx);
    }
    
    public boolean isValidBST(TreeNode root) {
        return validateBST(root , Long.MIN_VALUE , Long.MAX_VALUE);
    }
}
